api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -944,6 +920,9 @@ message Cluster {
   // [#next-major-version: make this a list of typed extensions.]
   map<string, google.protobuf.Any> typed_extension_protocol_options = 36;
 
+  // [#extension-category: envoy.clusters.dns]
+  envoy.extensions.clusters.dns.v3.DnsConfig dns_config = 59;
```

Comment:
We don't need a new extension point here.  The DNS cluster config should instead be loaded via the existing `cluster_type` field:

https://github.com/envoyproxy/envoy/blob/1173629e531abf758f011c2da15da739f72881c6/api/envoy/config/cluster/v3/cluster.proto#L833

Reply:
@markdroth Oh I see, so we are adding relevant DNS fields to the cluster config types of each cluster type. So I would add a `api/envoy/extensions/clusters/strict_dns/v3/cluster.proto` file with the corresponding fields. and same for logical_dns, etc

Reply:
That would work.  But it may also be the case that there's enough in common (config-wise and/or implementation-wise) between STRICT_DNS and LOGICAL_DNS that we can combine them into a single DNS cluster type, and the config for that DNS cluster type can have a field that says whether it uses STRICT or LOGICAL behavior.

Reply:
@markdroth wait... So would I add an extension point to the typed config. So the configuration would like

```proto
  clusters:
  - name: my_cluster
    ...
    cluster_type:
      name: envoy.clusters.strict_dns
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.clusters.strict_dns.v3.ClusterConfig
      dns_config:
         dns_jitter: false
         ... the rest of the dns_* fields
```

Reply:
The new DNS cluster type should not use any of the existing `dns_*` fields in the `Cluster` message; those fields should all be copied into the new `envoy.extensions.clusters.dns_cluster.v3.DnsCluster` message.

The existing `dns_*` fields in the `Cluster` message should be used only for backward-compatibility reasons, for existing `Cluster` resources that use `STRICT_DNS` or `LOGICAL_DNS` and configure these existing fields.  However, Envoy should internally convert such a config into the corresponding new-style config that uses the new `envoy.extensions.clusters.dns_cluster.v3.DnsCluster` message.

Reply:
@markdroth So to the new way of configuring a strict dns cluster would be

```
  clusters:
  - name: my_cluster
    ...
    cluster_type:
      name: envoy.clusters.strict_dns
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.clusters.dns_cluster.v3.DnsCluster
         dns_jitter: false
         ... the rest of the dns_* fields
```

I understand that there is a new and old style config and that we need to keep the old style for backcompatibility, but internally we want to use the new style.
I just don't understand where the new extension point should be mounted

Reply:
In the above example, I just worry that we are forcing both strict and logical dns to share the same typed_cofig

Reply:
> In the above example, I just worry that we are forcing both strict and logical dns to share the same typed_cofig

Why would that a problem, as long as there's still a way to configure the behavior for each?  My suspicion is that if we had separate config messages for STRICT_DNS and LOGICAL_DNS, then both messages would have almost identical sets of fields.  If that's the case, then the duplication doesn't really make sense, which is why I suggested having a single message that can be used for both with a knob that controls which behavior is used.

The type present in the `type_url` field dictates which cluster implementation gets used, so if we have two different implementations for STRICT_DNS and LOGICAL_DNS, then we would need probably-identical config messages, like this:

```
message StrictDnsCluster {
  // ...a bunch of DNS fields...
}

message LogicalDnsCluster {
  // ...exactly the same set of DNS fields...
}
```

In contrast, if we have a single cluster implementation for both DNS types with a knob to control the behavior, we can have a single config proto, like this:

```
message DnsCluster {
  // if true, STRICT_DNS semantics; otherwise, LOGICAL_DNS semantics
  // (could also be an enum or whatever)
  bool strict = N;

  // ...the set of DNS fields...
}
```

Reply:
@markdroth I agree that most fields are Dns-related and would thus be duplicated, but I was thinking something along the lines of

```
message DnsConfig {
   // ... DNS fields
}

message StrictDnsCluster {
  DnsConfig dns_config = N
}

message LogicalDnsCluster {
  DnsConfig dns_config = N
}
```
This way we don't have duplicate code and 

1. In the future, if we want to add a logical dns-specific field, it won't appear in the strict dns configuration (same for strict dns).
2. It follows the general pattern (AFAIK) of each cluster type having its own config type. We see this pattern for filters too.

Let me know what you think. I'm still new to envoy so if you feel that having a shared config type is the way to go, I'm happy to implement it.

Reply:
My inclination is that we're unlikely to have too many fields that are specific to either STRICT_DNS or LOGICAL_DNS, so combining them makes the most sense.  But I'd be open to input from @wbpcode or other Envoy maintainers.